torch/distributed/algorithms/ddp_comm_hooks/default_hooks.py

```diff
@@ -65,3 +67,78 @@ def decompress(fut):
         return [decompressed_tensor]
 
     return fut.then(decompress)
+
+
+class AllReduceFirstKStepsState(object):
+    """
+    State information required by allreduce_first_k_steps_hook to run gradient averaging for the first K steps.
+    Need to specify ``k``.
+    """
+
+    __slots__ = [
+        "process_group",
+        "k",
+        "step",
+    ]
+
+    def __init__(
+        self,
+        process_group,
+        k,
+    ):
+        self.process_group = process_group
+        self.k = k
+        self.step = 0
+
+        logging.info(
+            "Allreduce-first-k-steps hook runs allreduce for the first {} steps.".format(
+                self.k,
+            )
+        )
+
+    def maybe_increase_step(self, bucket):
+        # Since bucket 0 is the last bucket to allreduce in an iteration.
+        # Only increase `step` when bucket 0 is processed.
+        if bucket.get_index() == 0:
+            self.step += 1
+
+        if self.step == self.k:
+            logging.info("No more gradient averaging after {} steps.".format(self.k))
+
+
+def allreduce_first_k_steps_hook(
+    state: AllReduceFirstKStepsState, bucket: dist._GradBucket
+) -> torch.futures.Future:
+    """
+    This DDP communication hook runs allreduce to average gradients for the first K steps,
+    and then stops any communication afterwards.
+    This hook is designed to support post-local SGD, by running gradient averaging before local SGD stage.
+    It is user's responsibility to perform model parameter averaging at local SGD stage after K steps.
+
+    Args:
+        state (AllReduceFirstKStepsState): State information to run allreduce for the first K steps.
+        bucket (dist._GradBucket): Bucket that stores a 1D flattened gradient tensor that batches multiple per-variable tensors.
+            Note that since DDP comm hook only supports single process single device mode at this time,
+            only exactly one tensor is stored in this bucket.
+
+    Returns:
+        Future handler of the communication, which updates the gradients in place.
+
+    Example::
+        state = AllReduceFirstKStepsState(process_group=process_group, k=1000)
+        >>> ddp_model.register_comm_hook(state, allreduce_first_k_steps_hook)
+    """
+    process_group = state.process_group
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    # The input tensors in the bucket only has a single tensor.
+    input_tensors = bucket.get_tensors()
+
+    # Run allreduce in the first K steps.
+    if state.step < state.k:
+        state.maybe_increase_step(bucket)
+        return _allreduce_fut(group_to_use, input_tensors[0])
+    else:
+        # Afterwards K steps, directly return the input tensors as the output future.
+        ret_fut = torch.futures.Future()
+        ret_fut.set_result(input_tensors)
+        return ret_fut
```

Comment:
Shouldn't we implement local SGD here where we now average gradients every H steps? I'm not sure why we leave this upto the user? It should be easy to implement that logic here right?

Reply:
Local SGD needs to average parameters instead of gradients, and parameter averaging should be executed after `optimizer.step()`, so actually it cannot be implemented in the context of DDP at this time.